### PR TITLE
separate the loading data in py-script and update the notebook  accor…

### PR DESCRIPTION
**Pull request by saharmahdie**
…dingly

created the pytorch_dataset.py and use it in data-pytorch-use-pytorch-dataset-function.ipynb later we can delete  data-pytorch.ipynb.

### CODE REVIEW

File: tutorial/3_loading_data/data-pytorch-use-pytorch-dataset-function.ipynb [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+    "- This will be used to iterate through the data in batches during training."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dataloader = get_dataloader(PATH_TO_DATASET)"
```
Comment (VRehnberg) [inline tutorial/3_loading_data/data-pytorch-use-pytorch-dataset-function.ipynb]:
```suggestion
    "from pytorch_dataset import TinyImageNetDataset\n"
    "dataset = TinyImageNetDataset(path_to_dataset)\n"
    "dataloader = DataLoader(dataset, ...)"
```
Perhaps something like this in 4 and 5. That way we expose the options to the DataLoader directly in the notebook.

File: tutorial/3_loading_data/pytorch_dataset.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        img_array = np.array(Image.open(BytesIO(self.zf.read(imgpath))))
+        if img_array.ndim < 3:
+            # Greyscale to RGB
+            img_array = np.repeat(img_array[..., np.newaxis], 3, -1)
+
+        img_tensor = torch.from_numpy(img_array).permute(2, 0, 1).float()
+
+        # Get label from filename
+        label = self.get_label(imgpath)
+        return img_tensor, label
```
Comment (VRehnberg) [inline tutorial/3_loading_data/pytorch_dataset.py]: https://github.com/saharmahdie/alvis-intro/blob/main/tutorial/3_loading_data/pytorch_dataset.py#L43-L87

If we focus on 4 and 5, this Dataset is probably all/most of what we need (but you can check, if you think some other data utility that is used in 4 and 5 should be moved).

**Review by VRehnberg — CHANGES_REQUESTED** [review]
3 can be kept as is I think. As the task in 3 is about the dataset it's good if we expose it in the notebook.

But, in 4 and 5 we can hide it in a python function to make it clearer what the focus is in those notebooks.

File: tutorial/3_loading_data/data-pytorch-use-pytorch-dataset-function.ipynb (hunk 1 above)
Comment (saharmahdie) [reply tutorial/3_loading_data/data-pytorch-use-pytorch-dataset-function.ipynb]: ok, let me check these sections code.

File: tutorial/4_checkpointing/checkpointing-pytorch-use-pytorch-dataset-class.ipynb [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+   ],
+   "source": [
+    "import zipfile\n",
+    "\n",
+    "path_to_dataset = '/mimer/NOBACKUP/Datasets/tiny-imagenet-200/tiny-imagenet-200.zip'\n",
+    "\n",
+    "with zipfile.ZipFile(path_to_dataset, 'r') as zf:\n",
+    "    print(zf.namelist())\n"
+   ]
+  },
```
Comment (VRehnberg) [inline tutorial/4_checkpointing/checkpointing-pytorch-use-pytorch-dataset-class.ipynb]: This cell can be removed.
```suggestion
```

File: tutorial/4_checkpointing/checkpointing-pytorch-use-pytorch-dataset-class.ipynb [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    "            # Accuracy calculation\n",
+    "            #_, predicted = torch.max(est, 1)\n",
+    "            #correct += (predicted == labels).sum().item()\n",
+    "            #total += labels.size(0)\n",
+    "        \n",
+    "        # Average loss and accuracy\n",
+    "        loss /= n_batches\n",
+    "        #accuracy = correct / total\n",
+    "\n",
+    "        return loss, acc\n"
```
Comment (VRehnberg) [inline tutorial/4_checkpointing/checkpointing-pytorch-use-pytorch-dataset-class.ipynb]:
```suggestion
    "            \n",
    "            # Accuracy calculation\n",
    "            predicted = est.argmax(1)\n",
    "            correct += (est.argmax(1) == labels).sum().item()\n",
    "            total += labels.size(0)\n",
    "        \n",
    "        # Average loss and accuracy\n",
    "        loss /= n_batches\n",
    "        ac = correct / total\n",
    "\n",
    "        return loss, acc\n"
```
Yes, you're right that the accuracy calculation here is a bit sloppy, you could do something like this instead.

File: tutorial/4_checkpointing/pytorch_dataset.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import zipfile
+import os
+from typing import Tuple
+import numpy as np
+import torch
+from PIL import Image
+
+# Custom Type Hints
+LoadedFromZip = Tuple[str, object]  
+DataPoint = Tuple[torch.FloatTensor, int]
```
Comment (VRehnberg) [inline tutorial/4_checkpointing/pytorch_dataset.py]:
```suggestion
```
Think these are unused now, so we can remove them.

**Review by VRehnberg — CHANGES_REQUESTED** [review]
Looks mostly good:
 - [x] had a few minor comments
 - [x] you can replace checkpointing-pytorch.ipynb with the new notebook, we don't need both
 - [x] you could remove cell outputs and submit a clean notebook
     1. Right-click somewhere inside the notebook.
     2. Select "Clear Outputs of All Cells".

File: tutorial/4_checkpointing/pytorch_dataset.py (hunk 5 above)
Comment (saharmahdie) [reply tutorial/4_checkpointing/pytorch_dataset.py]: done

File: tutorial/4_checkpointing/checkpointing-pytorch-use-pytorch-dataset-class.ipynb (hunk 4 above)
Comment (saharmahdie) [reply tutorial/4_checkpointing/checkpointing-pytorch-use-pytorch-dataset-class.ipynb]: done

File: tutorial/4_checkpointing/checkpointing-pytorch-use-pytorch-dataset-class.ipynb (hunk 3 above)
Comment (saharmahdie) [reply tutorial/4_checkpointing/checkpointing-pytorch-use-pytorch-dataset-class.ipynb]: done
